### ISSUE TITLE
Compatibility with Meteor

### ISSUE THREAD

**Issue by ghost** [ajv-validator/ajv#159]
I get this error:

`async schema referenced by sync schema`

with this code:

```
if(!ajv.validate(personSchema, personToInsert)){
    return ajv.errors;
};
```

I have no idea what this means.  This whole thing looks synchronous to me, and I don't know what an async schema is.

Note: this is all run in a MeteorJS fiber.

Any help would be appreciated.

**Comment by epoberezkin** [comment]
I'd like to see the schemas. The message means that the schema with $async keyword is referenced by the schema without $async keyword.

**Comment by ghost** [comment]
personSchema:

``` javascript
{
    "$schema": "http://json-schema.org/draft-04/schema#",
    "id": "/",
    "type": "object",
    "properties": {
        "first_name": {
            "id": "first_name",
            "type": "string"
        },
        "middle_name": {
            "id": "middle_name",
            "type": "null"
        },
        "last_name": {
            "id": "last_name",
            "type": "string"
        },
        "is_person": {
            "id": "is_person",
            "type": "boolean"
        },
        "title": {
            "id": "title",
            "type": "string"
        },
        "date_of_birth": {
            "id": "date_of_birth",
            "type": "string"
        },
        "addresses": {
            "id": "addresses",
            "type": "array",
            "items": {
                "id": "0",
                "type": "object",
                "properties": {
                    "type": {
                        "id": "type",
                        "type": "string"
                    },
                    "line_1": {
                        "id": "line_1",
                        "type": "string"
                    },
                    "line_2": {
                        "id": "line_2",
                        "type": "string"
                    },
                    "city": {
                        "id": "city",
                        "type": "string"
                    },
                    "state": {
                        "id": "state",
                        "type": "string"
                    },
                    "zip": {
                        "id": "zip",
                        "type": "string"
                    }
                },
                "additionalProperties": false
            },
            "additionalItems": false
        },
        "emails": {
            "id": "emails",
            "type": "array",
            "items": {
                "id": "0",
                "type": "object",
                "properties": {
                    "type": {
                        "id": "type",
                        "type": "string"
                    },
                    "email": {
                        "id": "email",
                        "type": "string"
                    }
                },
                "additionalProperties": false
            },
            "additionalItems": false
        },
        "phones": {
            "id": "phones",
            "type": "array",
            "items": {},
            "additionalItems": false
        }
    },
    "additionalProperties": false,
    "required": [
        "last_name"
    ]
}
```

``` javascript
personToInsert= {
    "first_name": "barry",
    "middle_name": null,
    "last_name": "bob",
    "is_person": true,
    "title":"CEO",
    "date_of_birth":"2012-12-12",
    "addresses":[{
        "type":"home",
        "line_1":"1818 Bobby Lane",
        "line_2":"Apt 444",
        "city":"Chicago",
        "state":"OH",
        "zip":"24235"
    }],
    "emails": [{
        "type":"home",
        "email": "barry@bob.com"
    }],
    "phones": []
}
```

**Comment by epoberezkin** [comment]
I don't think I can reproduce the issue: https://tonicdev.com/esp/5707dc7fed8bd01200bfb7a6

**Comment by ghost** [comment]
Man, it must have something to do with how meteor is including the library.  Thanks for taking the time to help though!

**Comment by nrotstan** [comment]
It's definitely a compatibility issue with Meteor. I have the same problem, and it works fine in a plain node.js app. I guess my question, though, is why it's trying to load any schemas at all if none are referenced? Is it the meta schema it's loading? I've even tried leaving out a $schema, but it makes no difference. I wonder if there's a way to work around the problem -- would love to use ajv.

**Comment by epoberezkin** [comment]
Could you make some minimal code that fails? The error message is not related to loading schemas, it is about asynchronous validation. But there is no async validation in this schema so I don't really know what it misses.

**Comment by nrotstan** [comment]
Here's a minimal Meteor project that demonstrates the problem: [https://github.com/nrotstan/meteor-ajv-incompatibility](https://github.com/nrotstan/meteor-ajv-incompatibility)

**Comment by epoberezkin** [comment]
Meteor extends Function.prototype with `async` method:

``` JavaScript
function (allowReuseOfCurrentFiber) {
  return MeteorPromise.async(this, allowReuseOfCurrentFiber);
}
```

Ajv uses own property of validation functions with the same name `async` to indicate that this function is asynchronous (returns promise) - it is used for asynchronous validation. So when Ajv compiles meta-schema it thinks that some referenced schema is asynchronous because of this.

**Comment by epoberezkin** [comment]
@nrotstan thanks for the failing code.

This particular issue is fixed in 3.8.9. There may be some other cases that would fail for the same reason - please reopen the issue if you notice it. In [4.0](https://github.com/epoberezkin/ajv/issues/151) I will rename `async` property of validation function to `$async`.

Thank you!

**Comment by nrotstan** [comment]
Thank you for the fix, @epoberezkin -- it's much appreciated!

I did run into one more issue with Meteor and ajv using 3.8.9: for invalid schemas, ajv.errors does not appear to be set (and if I pass true to validateSchema for the second parameter so that it throws an exception, I get "[Error: schema is invalid:No errors]"). It works fine on a plain node.js app. I've updated that meteor-ajv-incompatibility project to demonstrate the new issue. Thanks again!
